### PR TITLE
Rationalize installation, run tests via ctest

### PR DESCRIPTION
**Pull request by aradi**
## Pull request template

- [X] Requested `develop` as target branch
- [ ] ~~Attached test suite log file~~
- [X] Alerted reviewers if edits impact CMake or Makefile files

Rationalizeed installation, added testing via ctest.

* It only writes into the build folder, leaving the source folder unchanged (as it should be :wink:)
* It enables parallel test execution (via `ctest -jN`), each test writes in a different folder
* It reads the stdout of ChIMES (instead of `debug.dat`), so that test also work if `DEBUG=0` is set (as in production binaries)
* It now also builds (and installs) the dynamically loadable library (called `libchimescalc_dl.so`), which can be loaded via `ctypes` in Python.
* It removes the installation of unnecessary testing binaries which are of no interest for ChIMES consumers
* It installs one stand-alone binary (called `chimescalc`).
* The GitHub CI contains now a minimal test of all tested interfaces (cpp, c, fortran, fortran08).

Note: The Python wrapper is not installed and tested during the CMake build, as that would require some further changes possibly  affecting the make based testing procedure. That should be made in a future PR.

ALERT: It changes the CMake framework, but only by extending it in a backwards compatible way.

### CODE REVIEW

**Comment by rk-lindsey** [comment]
@aradi thanks for the PR! A few notes:

0. Please target the `develop` branch - it is now up to date with `main`

1. `./install.sh` fails:

> ```
> -- Build files have been written to: /Users/lindsey11/Desktop/test/build
> make[2]: *** No rule to make target `CMakeFiles/ChimesCalc.dir/depend'.  Stop.
> make[1]: *** [CMakeFiles/ChimesCalc.dir/all] Error 2
> make: *** [all] Error 2
> make[2]: *** No rule to make target `CMakeFiles/ChimesCalc.dir/depend'.  Stop.
> make[1]: *** [CMakeFiles/ChimesCalc.dir/all] Error 2
> make: *** [all] Error 2
> ```

2. `install.sh` (and by proxy, `CMakelists.txt`) also needs to have the same functionality as before, i.e. the same set of executables with the same names need to be produced and placed in `build`

3. I realize it may seem redundant, but `run_tests.sh` needs to work as before, i.e., successfully running `for compile in MAKEFILE  CMAKE` instead of just `for compile in MAKEFILE`

**Comment by aradi** [comment]
@rk-lindsey Thanks a lot for the comments!

I've rebased and targeted to `develop` now. Sorry, I did not realize, that this had been updated...

`./install.sh` works for me flawlessly, so I am wondering, why it fails for you. Do you have any further hints?

I was probably too enthusiastic in reorganizing your test strategy, without knowing too much about, why you have designed it, as it is, sorry... My plan was to take out CMake build from the original testing infrastructure provided by `run_tests.sh`, and make it more suitable for parallel testing and CI. `run_tests.sh` can still test the make-based build. And, I've just tested again, `run_tests.sh` works for me for the MAKE-based build without problems (provided I change the hard coded `python3.7` to my current python version :wink: ).

I see following problems with the current test system on `develop`:
- It writes into the source tree during testing. (That may be purely aesthetic, but this is nowadays an absolute no-go)
- It can not test single test cases but only all of them (or one fixed set, if SHORT is specified)
- Tests are run sequentially instead of in parallel, even if multiple threads are possible
- Failing tests are not generating non-zero exit codes, so the CI won't notice failure

Additionally, the current CMake build on `develop` has in my view following problems:
- It creates unnecessary libraries for Python. (Note: that we have merged all interfaces into one single library, therefore, it does not really make a sense to create a `serial` and a `direct` library, as both interfaces will be present in the same library.
- It creates executables and libraries with very generic (chimes-unspecific) names. That could be OK internally, but those libraries/executables (e.g. `lib-C_wrapper-direct_interface`) are installed during the installation. And even worse, they are installed to absolute non-standard locations. (The check `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` is not really helpful here, as during Conda-packaging, the PREFIX is overriden by Conda). So, basically the last changes in the CMake build on `develop` make a proper package generation via Conda absolutely impossible. :disappointed: 
- I think, installing a standalone binary (into the proper standardized location) which drives the library with file I/O makes indeed sense. But only one such binary should be installed and with a proper name (`e.g. chimescalc`).

I tried remedy all those problems with the commit above. The testing offered by CTest covers the same cases/APIs as the original one (apart of Python, see below). I understand, that it is not compatible with the current build and testing framework, but the current framework has several significant limitations, so staying compatible with it would make CI-testing and proper packaging impossible.

So, first, probably we should discuss, whether you can live with the fact, that the CMake based build differs from the Make based one in what it builds, how it builds and how it installs things, in order to provide a packaging and CI friendly build. If that's OK, we can try to improve on this PR a bit more and merge it at some point. But that would mean, that `run_test.sh` would not necessarily be able to run CMake based tests.

If that's a problem, feel free to close this PR. I'd then just create issues for things, which I think would be problematic for the Conda packaging and respect your decision on the testing framework. It is finally mostly you, who has to wait 4-40 times longer (depending on the nr. of cores you have available) when you do the testing. :wink: 

- A last note on Python testing: I did not have time to integrate that in the CTest, but I can do that later in this PR, provided we decide to go further with it.

**Comment by RKLindsey** [comment]
@aradi Thank you for the detailed response! Here are some thoughts:


> ./install.sh works for me flawlessly, so I am wondering, why it fails for you. Do you have any further hints?

I don't have any additional error messages, but I am testing on OSX (Catalina) and running `./install.sh` with all the default arguments. My C/C++ compilers are Clang, and Fortran compilers are gcc. Here's are my versions:

```
# Clang --version
Apple clang version 12.0.0 (clang-1200.0.32.21)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

#gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.21)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

> I see following problems with the current test system on develop:
>
> It writes into the source tree during testing. (That may be purely aesthetic, but this is nowadays an absolute no-go)

That's fine - `./run_tests.sh` can easily be updated to search for necessary files in build - I can take care of this later

> Tests are run sequentially instead of in parallel, even if multiple threads are possible

My plan was to modify `run_tests.sh` once the points in my prior comment were addressed. Specifically, I was going to have `run_tests.sh` auto-populate `FF`, `CFGS`, and `OPTIONS` from the `tests` file you introduced, so I don't have to update two separate files if additional tests are introduced. I was also going to have `run_tests.sh` use the `run_test_case` executable you used to actually launched the test. This would provide support for both testing approaches.

> Failing tests are not generating non-zero exit codes, so the CI won't notice failure

That can be easily remedied with the appropriate error function, e.g. `sys.exit(1)` as you've added in `run_test_case` - note that I am not at all opposed to supporting the `ctest` functionality you added!

> Additionally, the current CMake build on develop has in my view following problems:

Note: a lot of our decisions regarding compilation schemes, etc are motivated by a desire to provide extremely simple examples for all the different ways our APIs can be used, i.e., suitable for the users that have minimal, if any coding experience and therefore may not be familiar with best practices and subsequent implications for naming conventions, file placements, etc. This also means our documentation heavily references file/function naming, layout, etc. 

> It creates unnecessary libraries for Python. (Note: that we have merged all interfaces into one single library, therefore, it does not really make a sense to create a serial and a direct library, as both interfaces will be present in the same library.

This is an example of a decision made to demonstrate different usage cases

> It creates executables and libraries with very generic (chimes-unspecific) names. That could be OK internally, but those libraries/executables (e.g. lib-C_wrapper-direct_interface) are installed during the installation. 

See my very last response below.

> And even worse, they are installed to absolute non-standard locations. (The check CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT is not really helpful here, as during Conda-packaging, the PREFIX is overriden by Conda). 

Yes, I'm fine with removing that, so long as executables can be placed directly in build when `./install.sh` is used. Again, think of `install.sh` as a script to generate usage-case examples.

> I think, installing a standalone binary (into the proper standardized location) which drives the library with file I/O makes indeed sense. But only one such binary should be installed and with a proper name (e.g. chimescalc).

To be clear, our example cases (i.e. all the main files) are written for simplicity/readability, since they are intended as instructive implementation examples. They are therefore (intentionally) very inefficient and are not intended for production use in most scenarios (dftb+ is a special case because generally, system sizes will be tiny). Hence, the goal wasn't to provide a single binary for users to start using in their research, but rather a set of examples of how to _implement_ `chimesFF.{h,cpp}` and its various language APIs in their own, e.g., MD codes. The `serial_interface` is just showing them how they might do so. 

> I tried remedy all those problems with the commit above. The testing offered by CTest covers the same cases/APIs as the original one (apart of Python, see below). I understand, that it is not compatible with the current build and testing framework, but the current framework has several significant limitations, so staying compatible with it would make CI-testing and proper packaging impossible.

I understand. Again, I'm in favor of maintaining CTest compatibility/enabling CI-testing, I'd just like to do so in a way that minimally impacts the repo's architecture. Hopefully a few minor changes to `CMakeLists.txt` can satisfy both our needs!

> So, first, probably we should discuss, whether you can live with the fact, that the CMake based build differs from the Make based one in what it builds, how it builds and how it installs things, in order to provide a packaging and CI friendly build. If that's OK, we can try to improve on this PR a bit more and merge it at some point. But that would mean, that run_test.sh would not necessarily be able to run CMake based tests.

Right - from my end, I really just need `install.sh` to run without errors, and build the same number of executables directly in `build`. I can live with them having different names (e.g. `chimescalc-<original_name>`). Would that be enough to put develop back into compatibility with Conda/CTest/CI, or is creation of the multiple libraries/executables fundamentally incompatible?

**Comment by aradi** [comment]
@rk-lindsey Perfect, thanks for taking the time and sorry for breaking so many things without discussion in advance.

I think, I fixed the `install.sh` issue. It was likely due to my mistake to use identical names (apart of case convention) for two separate targets. Try 836193d, that should work.

Thinking once more about what you wrote and what I wish to achieve, I have following suggestions/remarks:

* For packaging, I think, we should build and install following binary components:
  -  ChIMES library (`libchimescalc.so` and/or `libchimescalc.a`)
  -  Fortran wrapper (`libchimescalc_fortran.so` and/or `libchimescalc_fortran.a)
  -  dynamically loadable library (I'd suggest to call it `libchimescalc_dl.so`, or something similar)
  -  a standalone executable (`chimescalc`) -- so that library users can verify their results with a tested, stand-alone tool as well (I as a library consumer found the existence of such an executable very useful :wink: ).

* I agree that, it is indeed a good idea to build further executables demonstrating the usage of the library in different scenarios. Additionally, those executables can be/are used to test the various interfaces. However, I don't think that it makes sense to install any of those additional"example"  binaries, as they offer the same functionality as the one standalone (`chimescalc` -- which is actually the C++ example) built/installed for the packaging. If, those additional demonstration binaries are not installed, their name can be actually arbitrary, so I am happy to revert them back to the original naming, if you wish.

* As for the two different libraries used or the demonstrating the Python capabilities: As now one single library can contain both interfaces, I think, consistently with how it is done for the normal library, one should build only one dynamically loadable library. There can be still two different Python examples, demonstrating the usage of the two different interfaces, but they both should load the same library.

* I think, it is important to think about the calculator as a distributed binary package, which people just use without achieving the source code. (As it should be the case for all proper libraries and programs. I know, that most atomic simulation tools do not fulfill that criteria, but that is only due to the ignorance of the scientists about best practices in informatics). Having different libraries for the different interfaces is rather confusing (also in LAPACK, you only have one single library for both, the drivers and the expert routines). Also, the documentation should demonstrate things according to the packaged/distributed structure, so that people downloading the package (e.g. via Conda) can apply it right away, and don't have to download the source or build binaries themselves in order to understand the explanations/references there.

* Finally, one could/should at some point also package up the Python interface (either as a Conda package or as pip package). I quite like the way, how the MBD library (which we also use) handles those things. but other solutions are also possible.

**Comment by RKLindsey** [comment]
@aradi Not a problem - again, your contributions have been extremely helpful. I think a lot of this just comes down to (at least initially) having two different goals for the code. I was hesitant about this being viewed as a standalone package because my typical usage case for `chimesFF.{cpp,h}` (i.e., MD with > million atoms) precludes viewing the(intentionally) inefficient `serial_interface` as a practical tool. Hence, I didn't want to give users the impression that it was is intended to be used in that manner.

> I think, I fixed the install.sh issue. It was likely due to my mistake to use identical names (apart of case convention) for two separate targets. Try 836193d, that should work.

Yes, that did it!

> Thinking once more about what you wrote and what I wish to achieve, I have following suggestions/remarks:
>
> For packaging, I think, we should build and install following binary components: (...)

I think your suggestions make sense - let's not have the extras built/installed for packaging. However, I would prefer to have a consistent naming convention, so don't worry about reverting them back.

> As for the two different libraries used or the demonstrating the Python capabilities: As now one single library can contain both interfaces, I think, consistently with how it is done for the normal library, one should build only one dynamically loadable library. There can be still two different Python examples, demonstrating the usage of the two different interfaces, but they both should load the same library.

That sounds reasonable.

> I think, it is important to think about the calculator as a distributed binary package, which people just use without achieving the source code. 

You make a good point. As I said above, I viewed `chimesFF.{cpp,h}` + APIs as the "product", not the inefficient `serial_interface` implementation. But, as you have rightfully pointed out, there are others that will actually want to use it.

> Also, the documentation should demonstrate things according to the packaged/distributed structure, so that people downloading the package (e.g. via Conda) can apply it right away, and don't have to download the source or build binaries themselves in order to understand the explanations/references there.

Excellent point. Adding an "Obtaining the code" > "Conda" page to the docs is a great idea. 

> Finally, one could/should at some point also package up the Python interface (either as a Conda package or as pip package). I quite like the way, how the MBD library (which we also use) handles those things. but other solutions are also possible.

Good suggestion - I'll add that to the to do list.

Finally, do you want to make the changes described in your first three main bullets in this PR or in a new one?

[tests.log](https://github.com/rk-lindsey/chimes_calculator/files/7747549/tests.log)

**Comment by aradi** [comment]
@rk-lindsey As far as I am concerned, I regard this PR as complete. It basically covers the first 3 points of the list, apart of the documentation (where it is probably much better anyway, if you change it than me :smile: ). Probably the only remaining issue is the names of the example binaries, which are built but not installed, as I changed them to `ŧest_*` in my ovIf you don't like it, I'd suggest to put a erwhelming enthusiasm... If you tell me, what name should be used there (or whether I should change back to the original ones), I can change accordingly. Accordingly, feel free to push into this branch (you have write access) a commit changing those.

Otherwise, I consider it as ready to be merged, but let me know, if it needs something else before it can be merged.

**Review by rk-lindsey — APPROVED** [review]

**Pull request merged**
